commit 02639abc505884f6e02cee315b37f6fc0213a808 (refs/changes/89/1189/1)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-12-28T16:36:25-08:00 (2 years, 9 months ago)
    
    joyent/node-cueball#66 want claim() to return in same event loop run if possible

diff --git a/lib/connection-fsm.js b/lib/connection-fsm.js
index cace7d3..0daab27 100644
--- a/lib/connection-fsm.js
+++ b/lib/connection-fsm.js
@@ -176,6 +176,15 @@ ConnectionFSM.prototype.claim = function (stack, cb) {
 		}
 	}
 	this.emit('claimAsserted');
+	/*
+	 * If we've transitioned to busy synchronously as a result of the
+	 * emit, we can return to our caller immediately. This saves an
+	 * an unnecessary delay going around the event loop again.
+	 */
+	if (this.isInState('busy')) {
+		self.removeListener('stateChanged', onStateChanged);
+		cb(null, self.cf_shadow, self.cf_conn);
+	}
 };
 
 /*
